File: backend/app/schemas/mapping.py
"""
Pydantic schemas for mapping operations.
"""
from datetime import datetime
from typing import Optional, Any, Dict
from pydantic import BaseModel, Field


class MappingBase(BaseModel):
    """Base schema for mapping."""
    source_field: str = Field(..., min_length=1, max_length=255)
    target_field: str = Field(..., min_length=1, max_length=255)
    transform_type: str = Field(default="1:1", max_length=50)
    transform_config: Optional[Dict[str, Any]] = None
    validation_rules: Optional[Dict[str, Any]] = None


class MappingCreate(MappingBase):
    """Schema for creating a mapping."""
    pass


class MappingUpdate(BaseModel):
    """Schema for updating a mapping."""
    source_field: Optional[str] = Field(None, min_length=1, max_length=255)
    target_field: Optional[str] = Field(None, min_length=1, max_length=255)
    transform_type: Optional[str] = Field(None, max_length=50)
    transform_config: Optional[Dict[str, Any]] = None
    validation_rules: Optional[Dict[str, Any]] = None


class MappingResponse(MappingBase):
    """Schema for mapping response."""
    id: int
    project_id: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

File: frontend/src/types/mapping.ts
/**
 * Type definitions for mapping operations
 */
import { Node, Edge } from 'reactflow';

export interface MappingResponse {
  id: number;
  project_id: number;
  source_field: string;
  target_field: string;
  transform_type: '1:1' | 'concat' | 'constant' | 'lookup' | 'split' | 'custom';
  transform_config: Record<string, any> | null;
  validation_rules: Record<string, any> | null;
  created_at: string;
  updated_at: string;
}

export interface MappingCreate {
  source_field: string;
  target_field: string;
  transform_type?: '1:1' | 'concat' | 'constant' | 'lookup' | 'split' | 'custom';
  transform_config?: Record<string, any>;
  validation_rules?: Record<string, any>;
}

export interface SchemaFieldNode {
  id: string;
  fieldName: string;
  dataType: string;
  sampleValues?: any[];
  nullCount?: number;
  uniqueCount?: number;
  side: 'source' | 'target';
  fileId: number;
}

export type MappingNode = Node<SchemaFieldNode>;
export type MappingEdge = Edge;
